### PR TITLE
Add level cache

### PR DESCRIPTION
**Pull request by godchen0212**
Add level cache
Issue #15540 

Signed-off-by: godchen0212 <qingxiang.chen@zilliz.com>

### CODE REVIEW

File: internal/querynode/query_collection.go [hunk 1]
```diff
@@ -1262,11 +1262,15 @@ func (q *queryCollection) retrieve(msg queryMsg) error {
 		if q.remoteChunkManager == nil {
 			return fmt.Errorf("can not create vector chunk manager for remote chunk manager is nil, msgID = %d", retrieveMsg.ID())
 		}
-		q.vectorChunkManager, err = storage.NewVectorChunkManager(q.localChunkManager, q.remoteChunkManager,
+		q.vectorChunkManager, err = storage.NewVectorChunkManager(q.remoteChunkManager,
```
Comment (xiaofan-luan) [inline internal/querynode/query_collection.go:1265]: where did localChunkManager used?

File: internal/util/cache/lru_cache.go [hunk 2]
```diff
@@ -90,35 +114,38 @@ func (c *LRU) evictedWorker() {
...
+	for c.size > c.capacity {
 		c.stats.evictedCount++
 		oldestE := c.evictList.Back()
 		if oldestE != nil {
 			c.evictList.Remove(oldestE)
-			kv := oldestE.Value.(*entry)
-			delete(c.items, kv.key)
-			if c.onEvicted != nil {
+			kv := oldestE.Value.(*Entry)
+			c.size -= uint64(kv.Size)
```
Comment (xiaofan-luan) [inline internal/util/cache/lru_cache.go:146]: why not use  c.storeHelper.MeasureSize?

File: internal/util/cache/lru_cache.go [hunk 3]
```diff
@@ -143,9 +170,10 @@ func (c *LRU) Remove(key Key) {
 	defer c.m.Unlock()
 	if e, ok := c.items[key]; ok {
 		c.evictList.Remove(e)
-		kv := e.Value.(*entry)
-		delete(c.items, kv.key)
-		if c.onEvicted != nil {
+		kv := e.Value.(*Entry)
+		delete(c.items, kv.Key)
+		c.size -= uint64(kv.Size)
```
Comment (xiaofan-luan) [inline internal/util/cache/lru_cache.go:175]: c.storeHelper.MeasureSize

File: internal/util/cache/lru_cache.go [hunk 4]
```diff
@@ -164,49 +192,51 @@ func (c *LRU) Keys() []Key {
...
 	return c.capacity
 }
 
 func (c *LRU) Purge() {
 	c.m.Lock()
 	defer c.m.Unlock()
 	for k, v := range c.items {
-		if c.onEvicted != nil {
-			c.evictedCh <- v.Value.(*entry)
+		if c.storeHelper.OnEvicted != nil {
```
Comment (xiaofan-luan) [inline internal/util/cache/lru_cache.go:215]: should maintain size?

File: internal/util/cache/lru_cache.go [hunk 5]
```diff
@@ -164,49 +192,51 @@ func (c *LRU) Keys() []Key {
...
+	for c.size > capacity {
 		oldestE := c.evictList.Back()
 		if oldestE != nil {
 			c.evictList.Remove(oldestE)
-			kv := oldestE.Value.(*entry)
-			delete(c.items, kv.key)
-			if c.onEvicted != nil {
+			kv := oldestE.Value.(*Entry)
+			delete(c.items, kv.Key)
+			c.size -= uint64(kv.Size)
```
Comment (xiaofan-luan) [inline internal/util/cache/lru_cache.go:238]: same here

File: internal/util/cache/lru_cache.go [hunk 6]
```diff
@@ -164,49 +192,51 @@ func (c *LRU) Keys() []Key {
...
+	return c.size
 }
 
-func (c *LRU) Capacity() int {
+func (c *LRU) Capacity() uint64 {
 	return c.capacity
 }
 
 func (c *LRU) Purge() {
 	c.m.Lock()
```
Comment (xiaofan-luan) [inline internal/util/cache/lru_cache.go:212]: How do we handle if evictedCh full? 
If purge we should handle evicthere but not send into evictedCh?

File: internal/storage/vector_chunk_manager.go [hunk 7]
```diff
@@ -140,65 +114,30 @@ func (vcm *VectorChunkManager) Exist(filePath string) bool {
...
-	return results, nil
-}
-
 // Read reads the pure vector data. If cached, it reads from local.
 func (vcm *VectorChunkManager) Read(filePath string) ([]byte, error) {
-	if vcm.cacheEnable {
-		if r, ok := vcm.cache.Get(filePath); ok {
-			at := r.(*mmap.ReaderAt)
+	if v, ok := vcm.cache.Get(filePath); ok {
+		if vcm.cache.localCacheEnable {
```
Comment (xiaofan-luan) [inline internal/storage/vector_chunk_manager.go:120]: why do we need localCacheEnable here?
That makes everything complicated

File: internal/storage/level_cache.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	memCache         *cache.LRU
+	localFileCache   *cache.LRU
+	localCacheEnable bool
+
+	memoryHitsCount     int
+	localCacheHitsCount int
+}
+
+func (l *LevelCache) Add(key cache.Key, value cache.Value) {
+	if l.localCacheEnable {
```
Comment (xiaofan-luan) [inline internal/storage/level_cache.go:21]: If it's level cache, why does filecache and memcache has same value.
File cache should be filled only when memory cache evict

File: internal/storage/level_cache.go [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+			l.localCacheHitsCount++
+			l.memCache.Add(key, v)
+			return v, ok
+		}
+	}
+	return nil, false
+}
+
+func (l *LevelCache) Remove(key cache.Key) {
+	l.memCache.Remove(key)
```
Comment (xiaofan-luan) [inline internal/storage/level_cache.go:47]: if mem cache in , then no need to move from localFileCache

File: internal/util/paramtable/component_param.go [hunk 10]
```diff
@@ -748,21 +752,17 @@ func (p *queryNodeConfig) initOverloadedMemoryThresholdPercentage() {
 	p.OverloadedMemoryThresholdPercentage = float64(thresholdPercentage) / 100
 }
 
+func (p *queryNodeConfig) initLocalFileCacheLimit() {
+	p.LocalFileCacheLimit = p.Base.ParseUint64WithDefault("queryNode.cache.localFileCacheLimit", 2147483648)
```
Comment (xiaofan-luan) [inline internal/util/paramtable/component_param.go:756]: localFileCacheLimit should be far larger than memoryLimit.
memoryLimit -> 256MB
localFileCacheLimit -> 16GB

File: internal/util/cache/lru_cache.go (hunk 2 above)
Comment (godchen0212) [reply internal/util/cache/lru_cache.go:146]: For measureSize may produce overhead, I hope it is used for once. So size has been assign into Entry in line 135.

File: internal/util/cache/lru_cache.go (hunk 3 above)
Comment (godchen0212) [reply internal/util/cache/lru_cache.go:175]: For measureSize may produce overhead, I hope it is used for once. So size has been assign into Entry in line 135.

File: internal/util/cache/lru_cache.go (hunk 5 above)
Comment (godchen0212) [reply internal/util/cache/lru_cache.go:238]: For measureSize may produce overhead, I hope it is used for once. So size has been assign into Entry in line 135.

File: internal/storage/vector_chunk_manager.go (hunk 7 above)
Comment (godchen0212) [reply internal/storage/vector_chunk_manager.go:120]: If localFileCache is enabled, we can mmap file into memory. We cache *mmap.ReaderAt in memory.
If localFileCache is not enabled, we cache byte array into memory.
The value v may be []byte or *mmap.ReadAt, so the complexity is unavoidable.

File: internal/util/cache/lru_cache.go (hunk 6 above)
Comment (godchen0212) [reply internal/util/cache/lru_cache.go:212]: lgtm

File: internal/storage/level_cache.go (hunk 9 above)
Comment (godchen0212) [reply internal/storage/level_cache.go:47]: lgtm

**Comment by xiaofan-luan** [comment]
Hey @godchen0212, are we still working on this issue?

**Comment by xiaofan-luan** [comment]
I will take the cache implementation

**Pull request closed without merging**